addons/bus/static/src/crosstab_bus.js

```diff
@@ -180,111 +149,21 @@ export class CrossTab extends Longpolling {
-                }
-            }
-
-            if (heartbeatValue !== this.lastHeartbeat) {
-                // someone else is also master...
-                // it should not happen, except in some race condition situation.
-                this._isMasterTab = false;
-                this.lastHeartbeat = 0;
-                peers[this._id] = now;
-                this._callLocalStorage('setItem', 'peers', peers);
-                this.stopPolling();
-                this.trigger('no_longer_master');
-            } else {
-                this.lastHeartbeat = now;
-                this._callLocalStorage('setItem', 'heartbeat', now);
-                this._callLocalStorage('setItem', 'peers', cleanedPeers);
-            }
-        } else {
-            //update own heartbeat
-            peers[this._id] = now;
-            this._callLocalStorage('setItem', 'peers', peers);
-        }
-
-        var hbPeriod = this._isMasterTab ? this.MASTER_TAB_HEARTBEAT_PERIOD : this.TAB_HEARTBEAT_PERIOD;
-        this._heartbeatTimeout = browser.setTimeout(this._heartbeat.bind(this), hbPeriod);
-    }
-
     /**
      * @private
      */
     _registerWindowUnload() {
         browser.addEventListener('unload', this._onUnload.bind(this));
     }
 
-    /**
-     * Check with the local storage if the current tab is the master tab.
-     * If this tab became the master, trigger 'become_master' event
-     *
-     * @private
-     */
-    _startElection() {
-        if (this._isMasterTab) {
-            return;
-        }
-        //check who's next
-        var now = new Date().getTime();
-        var peers = this._callLocalStorage('getItem', 'peers', {});
-        var heartbeatKillOld = now - this.HEARTBEAT_KILL_OLD_PERIOD;
-        var newMaster;
-        for (var peerName in peers) {
-            //check for dead peers
-            if (peers[peerName] < heartbeatKillOld) {
-                continue;
-            }
-            newMaster = peerName;
-            break;
-        }
-
-        if (newMaster === this._id) {
-            //we're next in queue. Electing as master
-            this.lastHeartbeat = now;
-            this._callLocalStorage('setItem', 'heartbeat', this.lastHeartbeat);
-            this._callLocalStorage('setItem', 'master', true);
-            this._isMasterTab = true;
-            this.startPolling();
-            this.trigger('become_master');
-
-            //removing master peer from queue
-            delete peers[newMaster];
-            this._callLocalStorage('setItem', 'peers', peers);
-        }
-    }
-
     /**
      * Update localstorage channels of with the channels of this tab.
      *
      * @private
      * @return {boolean} true if the aggregated channels has changed.
      */
     _updateChannels() {
-        const currentPeerIds = new Set(Object.keys(this._callLocalStorage('getItem', 'peers') || {}));
+        const currentPeerIds = new Set(this.env.services['multiTab'].getAllTabIds());
```

Comment:
this looks like a leak of abstraction. if you move the multi tab handling out of the crosstab bus, then the "multi tab handling" concern should be properly factored out of this file. it looks like what you want to do is save and retrieve some data associated to a tab. Then the multi tab service should be designed to allow that.